certbot/auth_handler.py

```diff
@@ -47,242 +41,142 @@ def __init__(self, auth, acme_client, account, pref_challs):
 
-                # Send all Responses - this modifies achalls
-                self._respond(aauthzrs, resp, best_effort)
-
-        # Just make sure all decisions are complete.
-        self.verify_authzr_complete(aauthzrs)
+            # Inform the ACME CA server that challenges are available for validation.
+            for achall, resp in zip(achalls, resps):
+                self.acme.answer_challenge(achall.challb, resp)
 
-        # Only return valid authorizations
-        ret_val = [aauthzr.authzr for aauthzr in aauthzrs
-                   if aauthzr.authzr.body.status == messages.STATUS_VALID]
+            # Wait for authorizations to be checked.
+            self._poll_authorizations(authzrs, max_retries, best_effort)
 
-        if not ret_val:
-            raise errors.AuthorizationError(
-                "Challenges failed for all domains")
+            # Keep validated authorizations only. If there is none, no certificate can be issued.
+            authzrs_validated = [authzr for authzr in authzrs
+                                 if authzr.body.status == messages.STATUS_VALID]
+            if not authzrs_validated:
+                raise errors.AuthorizationError('All challenges have failed.')
 
-        return ret_val
+            return authzrs_validated
 
-    def _choose_challenges(self, aauthzrs):
+    def _poll_authorizations(self, authzrs, max_retries, best_effort):
+        """
+        Poll the ACME CA server, to wait for confirmation that authorizations have their challenges
+        all verified. The poll may occur several times, until all authorizations are checked
+        (valid or invalid), or after a maximum of retries.
+        """
+        authzrs_to_check = {index: (authzr, None)
+                            for index, authzr in enumerate(authzrs)}
+        # Give an initial second to the ACME CA server to check the authorizations
+        sleep_seconds = 1
+        for _ in range(max_retries):
+            # Wait for appropriate time (from Retry-After, initial wait, or no wait)
+            if sleep_seconds > 0:
+                time.sleep(sleep_seconds)
+            # Poll all updated authorizations.
+            authzrs_to_check = {index: self.acme.poll(authzr) for index, (authzr, _)
+                                in authzrs_to_check.items()}
+            # Update the original list of authzr with the updated authzrs from server.
+            for index, (authzr, _) in authzrs_to_check.items():
+                authzrs[index] = authzr
+
+            # Handle failed authorizations: with best effort this is only a warning,
+            # otherwise an exception.
+            authzrs_failed = [authzr for authzr, _ in authzrs_to_check.values()
+                              if authzr.body.status == messages.STATUS_INVALID]
+            if authzrs_failed:
+                # Display report to the user
+                _report_failed_authzrs(authzrs_failed, self.account.key)
+                if not best_effort:
```

Comment:
I think this is a useful addition to Certbot's output when `best_effort` is used.

One change being made here which I didn't notice before is that if Certbot is run in `--quiet` mode and you fail a nonempty subset of challenges with `--allow-subset-of-names`, on `master` you see:
```
Challenge failed for domain example.org
```
while on this branch nothing about the failure is printed to the terminal.

Can we add this message back? It might be worth logging the message unconditionally so people see it when `best_effort` isn't used as well.

Reply:
It is not exactly the same logic than before. The warning would be superfluous in two situations:
* without best effort, because in this case, you have an exception already
* with best effort without quiet, because you have a way better report just after

So it is useful only for quiet mode with best_effort=true. But if you ask for quiet, and choose an option where challenges failures are not critical, why would we display theses warnings ?

Reply:
Despite it being the current behavior, you could certainly make an argument that these warnings shouldn't be displayed when using best_effort in quiet mode.

I think for me what feels like "correct" behavior depends on how you're using `--allow-subset-of-names`. The initial motivation behind the flag was damage mitigation. If you're running a server hosting multiple domains and you're failing challenges for some reason, obtain an updated cert for as many domains as possible rather than letting it expire for all domains in the cert.

If you're using the flag in this way, I think I'd expect to be clearly told when domains are being dropped. If, however, you're using it because you want to remove a domain from a cert, the warning is probably unnecessary. (I might argue that doing the latter is a bad way to do things though because you may drop more domains than you expect. In this case, explicitly saying what domains were dropped may still be helpful).

I'm certainly biased, but I personally prefer we continue to print the warning when using `--allow-subset-of-names`. I'm not that worried about it though so if you disagree, we can leave it out.

Reply:
I see your point. `--allow-subset-of-names` is more a mitigation than a feature. And `quiet` does not mean `silent`. It makes sense, in the context of a mitigation, to inform the user about the fact that this mitigation has been triggered.

I update the whole process of reporting like this:
* during the loop, log warn for each failed authzr received
* at the end of the loop, build the consolidated report, and show it except if `quiet`
* then if best_effort is not `true`, raise an exception about the fact that some challenges are failed

Then with this I think we provide sufficient information in all scenarios.